### ISSUE TITLE
investigate whether garbage production can be minimized

### ISSUE THREAD

**Issue by mhuesch** [sacredcapital/rep_lang#49]
migrating a todo from #40 to an issue here:

    diagnose cause of duplicated entries in sto_vec and ideally remove them

[these two match blocks](https://github.com/sacredcapital/rep_lang/blob/be7982d5e0695f8423088852067b318ae5a770d1/rep_lang_runtime/src/eval.rs#L140-L156) create some of the extra garbage. I am not sure how to eliminate them at present, though.

they represent the notion that "interpretation of a thunk produces intermediate store allocations, including the allocation of the result of evaluation, which is itself stored in the store. thus at the end, we perform a copy of the result, setting the unevaluated thunk to an evaluated thunk which wraps the value.

## possible ideas

the approach demonstrated in https://github.com/sacredcapital/rep_lang/tree/sto_cell_redirect seems quite viable and like it would close this issue.

**Comment by mhuesch** [comment]
example of redirects in action:

```
vayu :: ~/projects/rep_lang ‹laziness-post-sto*› » cargo run --bin rle ./rep_lang_concrete_syntax/examples/stdlib.rl | grep CellRedirect | wc -l
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/rle ./rep_lang_concrete_syntax/examples/stdlib.rl`
180
vayu :: ~/projects/rep_lang ‹laziness-post-sto*› » cargo run --bin rle ./rep_lang_concrete_syntax/examples/stdlib.rl | grep CellThunk | wc -l
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/rle ./rep_lang_concrete_syntax/examples/stdlib.rl`
224
git push --set-upstream origin sto_cell_redirect
```

`rle` currently prints a big debug blob showing all store elements. we filter them to see which are thunks and which are redirects.

`180` are redirects and `224` are actual thunks, out a total of `404`. not all values are the same size, and we will still allocate points, so it doesn't represent quite the 0.45% savings suggested. but it might be somewhere close... 😀

**Comment by mhuesch** [comment]
closed via #50.
